### PR TITLE
Share a new rgb matrix keyboard --Skyloong/dt400

### PR DESCRIPTION
**Pull request by JZ-Skyloong**
## Description

Share a new qmk keyboard skyloong/dt40.  The keymap and rgb Matrix effects are compatible with qmk.

## Types of Changes

- [ ] Core
- [ ] Bugfix
- [ ] New feature
- [ ] Enhancement/optimization
- [x] Keyboard (addition or update)
- [ ] Keymap/layout/userspace (addition or update)
- [ ] Documentation

## Issues Fixed or Closed by This PR

* 

## Checklist

- [x] My code follows the code style of this project: [**C**](https://docs.qmk.fm/#/coding_conventions_c), [**Python**](https://docs.qmk.fm/#/coding_conventions_python)
- [x] I have read the [**PR Checklist** document](https://docs.qmk.fm/#/pr_checklist) and have made the appropriate changes.
- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [x] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [x] I have added tests to cover my changes.
- [x] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

### CODE REVIEW

File: .vscode/settings.json [hunk 1]
```diff
@@ -30,5 +30,19 @@
...
+           "path": "D:/QMK_MSYS/usr/bin/bash.exe",
+           "env": {
+               "MSYSTEM": "MINGW64",
+               "CHERE_INVOKING": "1"
+           },
+            "args": ["--login"]
+        }
+
+    },
+    "terminal.integrated.cursorStyle": "line"
```
Comment (zvecr) [inline .vscode/settings.json:47]: This can not be done as part of a keyboard PR
```suggestion
```

File: keyboards/skyloong/dt40/halconf.h [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+
+#pragma once
+
+#define HAL_USE_I2C TRUE
+#define HAL_USE_PWM TRUE
+#define HAL_USE_RTC TRUE
+//#define HAL_USE_SPI TRUE
+//#define SPI_USE_WAIT TRUE
+//#define SPI_SELECT_MODE SPI_SELECT_MODE_PAD
+
```
Comment (zvecr) [inline keyboards/skyloong/dt40/halconf.h:26]:
```suggestion
#define HAL_USE_PWM TRUE

```

File: keyboards/skyloong/dt40/info.json [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+        "command": false,
+        "console": false,
+        "extrakey": true,
+        "mousekey": true,
+        "nkro": true,
+        "rgb_matrix": true,
+        "rgblight":false
+    },
+
+
```
Comment (zvecr) [inline keyboards/skyloong/dt40/info.json:26]:
```suggestion
        "rgb_matrix": true
    },

```

File: keyboards/skyloong/dt40/keymaps/default/readme.md
Comment (zvecr) [inline keyboards/skyloong/dt40/keymaps/default/readme.md:1]: This file can be removed.

File: keyboards/skyloong/dt40/keymaps/via/keymap.c [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+    _LOWER,
+    _RAISE,
+    _ADJUST,
+};
+
+// Defines the keycodes used by our macros in process_record_user
+enum custom_keycodes {
+    QMKBEST = SAFE_RANGE,
+    QMKURL
+};
```
Comment (zvecr) [inline keyboards/skyloong/dt40/keymaps/via/keymap.c:32]:
```suggestion
```

File: keyboards/skyloong/dt40/mcuconf.h [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+#define STM32_I2C_USE_I2C1 TRUE
+
+//#undef STM32_SPI_USE_SPI2
+//#define STM32_SPI_USE_SPI2 FALSE
+
+#undef STM32_PWM_USE_TIM1
+#define STM32_PWM_USE_TIM1 TRUE
+
+#undef STM32_PWM_USE_ADVANCED
+#define STM32_PWM_USE_ADVANCED TRUE
```
Comment (zvecr) [inline keyboards/skyloong/dt40/mcuconf.h:36]:
```suggestion
```

File: keyboards/skyloong/dt40/rules.mk [hunk 6]
```diff
@@ -0,0 +1,12 @@
+
+# RGB matrix enabled
+//RGB_MATRIX_ENABLE = yes
+//RGBLIGHT_ENABLE = yes
+
+# select RGB matrix driver IC
+//RGB_MATRIX_DRIVER = WS2812
+//RGBLIGHT_DRIVER = WS2812
+
+# select driver type of WS2812
```
Comment (zvecr) [inline keyboards/skyloong/dt40/rules.mk:10]:
```suggestion
```

File: keyboards/skyloong/dt40/readme.md [hunk 7]
```diff
@@ -0,0 +1,82 @@
...
+ * T   ---->  Key " F5 " ;
+ * Y   ---->  Key " F6 " ;
+ * U   ---->  Key " F7 " ;
+ * I   ---->  Key " F8 " ;
+ * O   ---->  Key " F9 " ;
+ * P   ---->  Key " F10 " ;
+ * K   ---->  Key " F11 " ;
+ * L   ---->  Key " F12 " ;
+
+
```
Comment (zvecr) [inline keyboards/skyloong/dt40/readme.md:57]:
```suggestion
```

**Review by zvecr — CHANGES_REQUESTED** [review]

File: .vscode/settings.json [hunk 8]
```diff
@@ -30,5 +30,4 @@
     },
     "clangd.arguments": [
         "--header-insertion=never"
-    ]
 }
```
Comment (zvecr) [inline .vscode/settings.json:33]:
```suggestion
    ]
}
```

Comment (JZ-Skyloong) [reply .vscode/settings.json]: Thank you for your reminding and suggestion

**Review by JZ-Skyloong — COMMENTED** [review]
I agree with your suggestion.
 Thank you

File: keyboards/skyloong/dt40/dt40.c [hunk 9]
```diff
@@ -0,0 +1,60 @@
...
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+
+
+#include "quantum.h"
+
+#ifdef RGB_MATRIX_ENABLE
+#define RGB_MATRIX_LED_COUNT 44
```
Comment (drashna) [inline keyboards/skyloong/dt40/dt40.c:22]: This doesn't do anything here.
```suggestion
```

And the led config should be moved to info.json (`qmk info -kb skyloong/dt40 -f json` should help with that)

File: keyboards/skyloong/dt40/readme.md [hunk 10]
```diff
@@ -0,0 +1,34 @@
+# DT40
+
+![DT40](http://semitek.oss-cn-hangzhou.aliyuncs.com/qmk_imaging%2Fdt40%20layout05192023.jpg?OSSAccessKeyId=ogxQdo6geQXQVIK5&Expires=1684576837&Signature=GUNgvZ5GDL9uZY6CfpKYN1KvsJk%3D)
```
Comment (drashna) [inline keyboards/skyloong/dt40/readme.md:3]: please rehost on an image hosting site.

**Review by drashna — CHANGES_REQUESTED** [review]

File: keyboards/skyloong/dt40/readme.md (hunk 10 above)
Comment (JZ-Skyloong) [reply keyboards/skyloong/dt40/readme.md:3]: This host is our own OSS, which provides a faster and more stable connection, and we want to continue using it.

Comment (JZ-Skyloong) [reply keyboards/skyloong/dt40/readme.md]: OK,Rehosted it.

**Comment by JZ-Skyloong** [comment]
Update the file ws2812_pwm.c in platforms\chibios\drivers folder.

File: keyboards/skyloong/dt40/info.json [hunk 11]
```diff
@@ -0,0 +1,99 @@
...
+    "keyboard_name": "DT40",
+    "maintainer": "JZ-Skyloong",
+    "bootloader": "stm32duino",
+    "diode_direction": "ROW2COL",
+    "processor": "STM32F103",
+    "url": "https://github.com/JZ-Skyloong",
+    "usb": {
+        "device_version": "1.0.0",
+        "pid": "0x0040",
+        "vid": "0x1EA7"
```
Comment (drashna) [inline keyboards/skyloong/dt40/info.json:12]: This vendor ID is already in use, and should be changed. 
https://www.the-sz.com/products/usbid/index.php?v=0x1ea7&p=&n=

Ideally, it should be above 0x4000, as that hasn't been assigned yet.  And the higher, the better.

File: keyboards/skyloong/dt40/info.json [hunk 12]
```diff
@@ -0,0 +1,99 @@
...
+
+            ]
+        }
+    },
+
+
+
+  "rgb_matrix":{
+        "driver":"WS2812",
+        "pin": "B4",
```
Comment (drashna) [inline keyboards/skyloong/dt40/info.json:90]:
```suggestion
    "ws2812": {
        "pin": "B4"
    },
    "rgb_matrix":{
        "driver":"WS2812",
```

File: keyboards/skyloong/dt40/readme.md [hunk 13]
```diff
@@ -0,0 +1,34 @@
+# DT40
+
+![DT40](http://semitek.oss-cn-hangzhou.aliyuncs.com/qmk_imaging%2Fdt40%20layout05192023.jpg?OSSAccessKeyId=ogxQdo6geQXQVIK5&Expires=1720909578&Signature=nPWJ3jICUZ4XnLQSpk9MZBbaYuY%3D)
```
Comment (drashna) [inline keyboards/skyloong/dt40/readme.md:3]: please rehost image on Imgur.

File: platforms/chibios/drivers/ws2812_pwm.c [hunk 14]
```diff
@@ -308,7 +308,7 @@ void ws2812_init(void) {
     for (i = 0; i < WS2812_RESET_BIT_N; i++)
         ws2812_frame_buffer[i + WS2812_COLOR_BIT_N] = 0; // All reset bits are zero
 
-    palSetLineMode(RGB_DI_PIN, WS2812_OUTPUT_MODE);
+    palSetLineMode(WS2812_DI_PIN, WS2812_OUTPUT_MODE);
```
Comment (drashna) [inline platforms/chibios/drivers/ws2812_pwm.c:311]: please revert this change.

**Review by drashna — CHANGES_REQUESTED** [review]

File: keyboards/skyloong/dt40/info.json (hunk 11 above)
Comment (JZ-Skyloong) [reply keyboards/skyloong/dt40/info.json:12]: We are a mainland in china registered company of semitek and share the same VID. Our company has unified management of PID, which can ensure that ID is uniquely identified.

File: keyboards/skyloong/dt40/readme.md (hunk 13 above)
Comment (JZ-Skyloong) [reply keyboards/skyloong/dt40/readme.md]: OK, rehosted image on lmgur follow your suggest.

**Comment by JZ-Skyloong** [comment]
invaild? Why is it?
